src/FGS.Extensions.Hosting.Middleware/HostBuilderExtensions.cs

```diff
@@ -0,0 +1,49 @@
+using System;
+
+using FGS.Extensions.Hosting.Middleware.Abstractions;
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace FGS.Extensions.Hosting.Middleware
+{
+    /// <summary>
+    /// Extends <see cref="IHostBuilder"/> with functionality for configuring the usage of instances of <see cref="IHostingMiddleware"/> to intercept calls
+    /// to the later-built <see cref="IHost"/>.
+    /// </summary>
+    /// <example>
+    /// <code>
+    /// public static IWebHostBuilder CreateWebHostBuilder(string[] args) =&gt;
+    ///     WebHost.CreateDefaultBuilder(args)
+    ///         .UseStartup&lt;Startup&gt;()
+    ///         .UseHostingMiddleware&lt;MyHostingMiddleware&gt;();
+    /// </code>
+    /// </example>
+    public static class HostBuilderExtensions
+    {
+        /// <summary>
+        /// Adds hosting middleware functionality to the web host startup &amp; graceful shutdown process, relying on an instance of
+        /// <typeparamref name="THostingMiddleware"/> to provide interceptor behavior.
+        /// </summary>
+        /// <param name="hostBuilder">An instance of the program initialization abstraction that is to be modified.</param>
+        /// <param name="hostingMiddlewareFactory">Optionally creates or retrieves an instance of <typeparamref name="THostingMiddleware"/> to use.
+        /// If not specified, an instance will be requested from the <see cref="IServiceProvider"/>.</param>
+        /// <typeparam name="THostingMiddleware">The type of <see cref="IHostingMiddleware"/> that will be used to intercept web host startup &amp; graceful shutdown.</typeparam>
+        /// <returns>A program initialization abstraction that has been augmented to apply a <see cref="IHostingMiddleware"/> to the eventually-created <see cref="IHost"/>.</returns>
+        /// <remarks>Multiple calls to this will apply multiple layers of the hosting middleware stack, each one specific to that invocation and desired middleware.</remarks>
+        public static IHostBuilder UseHostingMiddleware<THostingMiddleware>(this IHostBuilder hostBuilder, Func<IServiceProvider, THostingMiddleware> hostingMiddlewareFactory = null)
+            where THostingMiddleware : IHostingMiddleware
```

Comment:
Perhaps I've forgotten something, but why do we need `THostingMiddleware` to be a generic type parameter instead of using `IHostingMiddleware` in its place in the following code? Obviously doing the latter would have ramifications on the few other places where this is used, just trying to get a grip on why we did this over the other?

Reply:
By using the type parameter, we more tightly constrain the invocation to `GetRequiredService` in the first statement of the function to only return the type of middleware that the user is attempting to wire up, rather than the latest-registered implementation of anything implementing `IHostingMiddleware`. The intent is that - if one has multiple middlewares - they will call `UseHostingMiddleware<…>` multiple times, once for each kind of middleware.

Reply:
Ahh, yup that's what I was overlooking! Thanks!